File: mifosng-android/src/instrumentTest/java/com/mifos/mifosxdroid/tests/CreateNewClientFragmentTest.java
package com.mifos.mifosxdroid.tests;

import android.view.KeyEvent;
import android.widget.Spinner;

import androidx.test.core.app.ActivityScenario;
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner;
import androidx.test.rule.ActivityTestRule;

import com.mifos.mifosxdroid.R;
import com.mifos.mifosxdroid.online.DashboardActivity;
import com.mifos.mifosxdroid.online.createnewclient.CreateNewClientFragment;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.text.SimpleDateFormat;
import java.util.Date;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.closeSoftKeyboard;
import static androidx.test.espresso.action.ViewActions.scrollTo;
import static androidx.test.espresso.action.ViewActions.typeText;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static androidx.test.platform.app.InstrumentationRegistry.getInstrumentation;
import static junit.framework.TestCase.assertNotNull;
import static junit.framework.TestCase.assertTrue;

@RunWith(AndroidJUnit4ClassRunner.class)
public class CreateNewClientFragmentTest {

    private ActivityScenario activityScenario;
    @Rule
    public ActivityTestRule<DashboardActivity> activityTestRule =
            new ActivityTestRule<>(DashboardActivity.class);

    @Before
    public void setUp() throws Exception {
        activityScenario = ActivityScenario.launch(DashboardActivity.class);
        openCreateClient();
    }

    @Test
    public void testViewsAreVisible() throws InterruptedException {

        Thread.sleep(3000);

        onView(withId(R.id.et_client_first_name))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.et_client_middle_name))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.et_client_last_name))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.et_client_mobile_no))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.et_client_external_id))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.sp_gender))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.tv_dateofbirth))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.sp_client_type))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.sp_client_classification))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.sp_offices))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.sp_staff))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.btn_submit))
                .perform(scrollTo())
                .check(matches(isDisplayed()));

        onView(withId(R.id.cb_client_active_status))
                .perform(scrollTo(), click())
                .check(matches(isDisplayed()));
        onView(withId(R.id.tv_submission_date))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
    }

    @Test
    public void testViewsAreNotNull() {
        assertNotNull(onView(withId(R.id.et_client_first_name)));
        assertNotNull(onView(withId(R.id.et_client_middle_name)));
        assertNotNull(onView(withId(R.id.et_client_last_name)));
        assertNotNull(onView(withId(R.id.et_client_mobile_no)));
        assertNotNull(onView(withId(R.id.et_client_external_id)));
        assertNotNull(onView(withId(R.id.sp_gender)));
        assertNotNull(onView(withId(R.id.tv_dateofbirth)));
        assertNotNull(onView(withId(R.id.sp_client_type)));
        assertNotNull(onView(withId(R.id.sp_client_classification)));
        assertNotNull(onView(withId(R.id.sp_offices)));
        assertNotNull(onView(withId(R.id.sp_staff)));
        assertNotNull(onView(withId(R.id.cb_client_active_status)));
        assertNotNull(onView(withId(R.id.tv_submission_date)));
        assertNotNull(onView(withId(R.id.btn_submit)));
    }

    @Test
    public void testOpenCreateClient() throws InterruptedException {
        CreateNewClientFragment createNewClientFragment
                = (CreateNewClientFragment) activityTestRule.getActivity()
                .getSupportFragmentManager().findFragmentById(R.id.container);

        assertNotNull(createNewClientFragment);
        assertTrue(createNewClientFragment.isAdded());
    }

    @Test
    public void testCreateClient() throws InterruptedException {

        Thread.sleep(3000);

        // First Name
        onView(withId(R.id.et_client_first_name))
                .perform(scrollTo())
                .perform(typeText("myFirstName"), closeSoftKeyboard());

        // Middle Name
        onView(withId(R.id.et_client_middle_name))
                .perform(scrollTo())
                .perform(typeText("myMiddleName"), closeSoftKeyboard());

        // Last Name
        onView(withId(R.id.et_client_last_name))
                .perform(scrollTo())
                .perform(typeText("myLastName"), closeSoftKeyboard());

        // Mobile Number
        onView(withId(R.id.et_client_mobile_no))
                .perform(scrollTo())
                .perform(typeText("987654321"), closeSoftKeyboard());

        // External Id
        onView(withId(R.id.et_client_external_id))
                .perform(scrollTo())
                .perform(typeText("123"), closeSoftKeyboard());

        // Gender
        Spinner spinnerGender = (Spinner) activityTestRule
                .getActivity().findViewById(R.id.sp_gender);
        String selectedGender = (String) spinnerGender.getAdapter().getItem(2);
        onView(withId(R.id.sp_gender))
                .perform(scrollTo(), click());
        onView(withText(selectedGender))
                .check(matches(isDisplayed()))
                .perform(click())
                .check(matches(isDisplayed()));

        // Client
        Spinner spinnerClient = (Spinner) activityTestRule
                .getActivity().findViewById(R.id.sp_client_type);
        String selectedClient = (String) spinnerClient.getAdapter().getItem(2);
        onView(withId(R.id.sp_client_type))
                .perform(scrollTo(), click());
        onView(withText(selectedClient))
                .check(matches(isDisplayed()))
                .perform(click())
                .check(matches(isDisplayed()));

        // Client Classification
        Spinner spinnerClientClassification = (Spinner) activityTestRule
                .getActivity().findViewById(R.id.sp_client_classification);
        String selectedClientClassification = (String) spinnerClientClassification
                .getAdapter().getItem(1);
        onView(withId(R.id.sp_client_classification))
                .perform(scrollTo(), click());
        onView(withText(selectedClientClassification))
                .check(matches(isDisplayed()))
                .perform(click())
                .check(matches(isDisplayed()));

        // Office
        Spinner spinnerOffice = (Spinner) activityTestRule
                .getActivity().findViewById(R.id.sp_offices);
        String selectedOffice = (String) spinnerOffice.getAdapter().getItem(0);
        onView(withId(R.id.sp_offices))
                .perform(scrollTo(), click());
        onView(withText(selectedOffice))
                .check(matches(isDisplayed()))
                .perform(click())
                .check(matches(isDisplayed()));

        // Staff
        Spinner spinnerStaff = (Spinner) activityTestRule
                .getActivity().findViewById(R.id.sp_staff);
        String selectedStaff = (String) spinnerStaff.getAdapter().getItem(0);
        onView(withId(R.id.sp_staff))
                .perform(scrollTo(), click());
        onView(withText(selectedStaff))
                .check(matches(isDisplayed()))
                .perform(click())
                .check(matches(isDisplayed()));

        // Activation Date
        onView(withId(R.id.cb_client_active_status))
                .perform(scrollTo(), click());
        SimpleDateFormat formatter = new SimpleDateFormat("dd-MM-yyyy");
        String currentDate = formatter.format(new Date());
        onView(withId(R.id.tv_submission_date))
                .check(matches(withText(currentDate)));

        // Submit
        onView(withId(R.id.btn_submit))
                .perform(scrollTo(), click());
    }

    private void openCreateClient() throws InterruptedException {
        getInstrumentation().sendKeyDownUpSync(KeyEvent.KEYCODE_MENU);
        onView(withText("Create Client"))
                .perform(click());
    }
}

File: mifosng-android/src/instrumentTest/java/com/mifos/mifosxdroid/tests/CreateNewClientFragmentTest.kt
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */
package com.mifos.mifosxdroid.tests

import android.test.ActivityInstrumentationTestCase2
import android.view.View
import com.mifos.mifosxdroid.R

/**
 * Created by ahmed fathy on 26/03/16.
 */
class CreateNewClientFragmentTest : ActivityInstrumentationTestCase2<DashboardActivity?>(
    DashboardActivity::class.java
) {
    /* fields */
    private var dashboardActivity: DashboardActivity? = null
    private var createNewClientFragment: CreateNewClientFragment? = null

    /* views */
    private var textViewDateOfBirth: TextView? = null
    private var textViewSubmissionDate: TextView? = null
    private var checkBoxActive: CheckBox? = null
    @Throws(Exception::class)
    fun setUp() {
        super.setUp()

        // open create client tab
        dashboardActivity = getActivity()
        dashboardActivity.runOnUiThread(Runnable { dashboardActivity.openCreateClient() })
        Thread.sleep(5000)

        // reference views
        createNewClientFragment = dashboardActivity
            .getSupportFragmentManager().findFragmentById(R.id.container) as CreateNewClientFragment
        textViewDateOfBirth = dashboardActivity.findViewById(R.id.tv_dateofbirth) as TextView
        textViewSubmissionDate = dashboardActivity.findViewById(R.id.tv_submission_date) as TextView
        checkBoxActive = dashboardActivity.findViewById(R.id.cb_client_active_status) as CheckBox
    }

    @SmallTest
    fun testViewsNotNull() {
        assertNotNull(textViewDateOfBirth)
        assertNotNull(textViewSubmissionDate)
        assertNotNull(checkBoxActive)
    }

    /**
     * - change birth date, check its text view is update
     * - change submission date, check its text view is updated and birth date isn't
     * - change birth date again, checks its text view is update, submission date isn't
     */
    @MediumTest
    @Throws(InterruptedException::class)
    fun testChangeDates() {

        // change birth date
        var birthDate = "01-01-1990"
        changeBirthDate(birthDate)
        assertEquals(textViewDateOfBirth.getText().toString(), birthDate)

        // check the active checkbox
        dashboardActivity.runOnUiThread(Runnable { checkBoxActive.setChecked(true) })
        Thread.sleep(1000)
        assertEquals(View.VISIBLE, textViewSubmissionDate.getVisibility())

        // change submission date
        val submissionDate = "02-02-2000"
        changeSubmissionDate(submissionDate)
        assertEquals(textViewSubmissionDate.getText().toString(), submissionDate)
        assertEquals(textViewDateOfBirth.getText().toString(), birthDate)

        // change date of birth again
        birthDate = "03-03-1989"
        changeBirthDate(birthDate)
        assertEquals(textViewSubmissionDate.getText().toString(), submissionDate)
        assertEquals(textViewDateOfBirth.getText().toString(), birthDate)
    }

    @Throws(InterruptedException::class)
    private fun changeBirthDate(dateOfBirth: String) {

        // open the date picker
        Espresso.onView(ViewMatchers.withId(R.id.tv_dateofbirth)).perform(ViewActions.scrollTo())
        Espresso.onView(ViewMatchers.withId(R.id.tv_dateofbirth)).perform(ViewActions.click())
        Thread.sleep(1000)
        assertTrue(createNewClientFragment.datePickerDateOfBirth.isAdded())

        // update date and close date picker
        dashboardActivity.runOnUiThread(Runnable { createNewClientFragment.onDatePicked(dateOfBirth) })
        createNewClientFragment.datePickerDateOfBirth.dismiss()
        Thread.sleep(1000)
    }

    @Throws(InterruptedException::class)
    private fun changeSubmissionDate(submissionDate: String) {
        // open the date picker
        Espresso.onView(ViewMatchers.withId(R.id.tv_submission_date))
            .perform(ViewActions.scrollTo())
        Espresso.onView(ViewMatchers.withId(R.id.tv_submission_date)).perform(ViewActions.click())
        Thread.sleep(1000)
        assertTrue(createNewClientFragment.datePickerSubmissionDate.isAdded())

        // update date and close date picker
        dashboardActivity.runOnUiThread(Runnable {
            createNewClientFragment.onDatePicked(
                submissionDate
            )
        })
        createNewClientFragment.datePickerSubmissionDate.dismiss()
        Thread.sleep(1000)
    }
}